packages/cmr-client/src/EarthdataLogin.ts

```diff
@@ -1,202 +1,155 @@
-   * The statuscode, statusmessage, error description, and error message are thrown and outputted.
-   *
-   * @throws {Error} - EarthdataLogin error
-   */
-  handleHttpError(error: Error) {
-    const statusCode = get(error, 'response.statusCode', error.code);
-    const statusMessage = get(error, 'response.statusMessage', error.message);
-    const responseErrorDescription = JSON.parse(get(error, 'response.body')).error_description;
-    const errorMessage = `EarthdataLogin error: ${responseErrorDescription},  statusCode: ${statusCode}, statusMessage: ${statusMessage}. Earthdata Login Request failed`;
+/**
+* This helper method is called in the functions that retrieve, create, and revoke EarthdataLogin
+* tokens for error-handling. If API call made to the EarthdataLogin endpoint results in an error.
+* The statuscode, statusmessage, error description, and error message are thrown and outputted.
+*
+* @returns {Error} - EarthdataLogin error
+*/
+const parseHttpError = (error: HTTPError, requestType: string): Error => {
+  const statusCode = error.response.statusCode;
+  const statusMessage = error.response.statusMessage || 'Unknown';
+  const errorBody = requestType === 'retrieve' || requestType === 'create' ? JSON.stringify(error.response.body) : error.response.body;
+  const message = `EarthdataLogin error: ${errorBody},  statusCode: ${statusCode}, statusMessage: ${statusMessage}. Earthdata Login Request failed`;
+  return new Error(message);
+};
 
-    throw new Error(errorMessage);
+/**
+  * The method for getting the token from the Earthdata Login endpoint. Sends a GET request
+  * with the users' base64 encoded username and password as a header for authorization. If the
+  * users' credentials are accepted the first unexpired token is retrieved, if one exists, and
+  * returned on their behalf, if not, an error is thrown. If the user does not have a token
+  * in Earthdata Login then undefined is returned to indicate to token needs to be created.
+  *
+  * @returns {string} the token
+*/
+export const retrieveEDLToken = async (username: string, password: string, edlEnv: string):
+Promise<string | undefined> => {
+  // response: get a token from the Earthdata login endpoint using credentials if exists
+  let rawResponse: any;
+  try {
+    rawResponse = await got.get(`${getEdlUrl(edlEnv)}/api/users/tokens`,
+      {
+        responseType: 'json',
+        username: username,
+        password: password,
+      });
+  } catch (error) {
+    if (error instanceof got.HTTPError) throw parseHttpError(error, 'retrieve');
+    throw parseCaughtError(error);
   }
+  const tokens = GetTokenResponseBody.parse(rawResponse.body);
+  const isTokenExpired = (token: Token) => new Date(token.expiration_date) < new Date();
+  const unExpiredTokens = tokens.filter((token: any) => !isTokenExpired(token));
+  return unExpiredTokens.length > 0 ? sortBy(unExpiredTokens, ['expiration_date'])[0].access_token : undefined;
```

Comment:
Assuming the test data is correct, and you are looking for the token that expires last, then this line is not going to give you the behavior you're expecting. For example:

```
>   sortBy(
...     [
...       { expiration_date: '1/1/2000' },
...       { expiration_date: '2/1/2001' },
...     ],
...     'expiration_date'
...   )
[ { expiration_date: '1/1/2000' }, { expiration_date: '2/1/2001' } ]
```

Reply:
No. The issue is that you're sorting the expiration dates as strings. With the test data that you're using, though, that is not going to sort them in date order.